### PR TITLE
Yi hdf5 reader

### PR DESCRIPTION
**Pull request by RA-Neuro**
Modified functions to allow reading of meghdf5 MEG data format generated by York Instruments systems as discussed with Robert Oostenveld 26/5/2021. Test data to be provided via email. Please let me know if there's anything I can do to improve this. 
Thanks,
Richard Aveyard

### CODE REVIEW

**Comment by robertoostenveld** [comment]
thanks, I downloaded the test file and will have a look at it.

**Comment by robertoostenveld** [comment]
Hi Richard, thanks for the code contribution. I have reviewed it here online, but prior to merging I will pull it and test it locally. I also plan to make some cosmetic changes to the code, and I will write some comments here to check those with you.

File: fileio/ft_filetype.m [hunk 1]
```diff
@@ -70,6 +70,7 @@
 %  - Neurodata Without Borders *.nwb
 %  - PhysioNet *.hea and *.dat
 %  - NIRx *.tpl, *.wl1 and *.wl2
+%  - York Instruments
```
Comment (robertoostenveld) [inline fileio/ft_filetype.m:73]: please add file extension

File: fileio/ft_filetype.m [hunk 2]
```diff
@@ -1539,6 +1540,11 @@
   type = 'duneuro_dgf';
   manufacturer = 'duneuro';
   content = 'geometrical meshes';
+  % known York Instruments file types
+elseif filetype_check_extension(filename, '.meghdf5')
+  type = 'York_Instruments_hdf5';
```
Comment (robertoostenveld) [inline fileio/ft_filetype.m:1545]: would you mind changing this into `yorkinstruments_hdf5`, i.e. a single string for the first part and only lower case? That nomenclature will make things easier for future extensions, such as adding a template layout for plotting (see https://www.fieldtriptoolbox.org/template/layout/)

File: fileio/ft_read_data.m [hunk 3]
```diff
@@ -1610,7 +1610,11 @@
       ft_hastoolbox('yokogawa', 1); % error if it cannot be added
       dat = read_yokogawa_data(filename, hdr, begsample, endsample, chanindx);
     end
-    
+
+  case 'York_Instruments_hdf5'
```
Comment (robertoostenveld) [inline fileio/ft_read_data.m:1614]: idem `yorkinstruments_hdf5`

File: fileio/ft_read_event.m [hunk 4]
```diff
@@ -2238,6 +2238,22 @@
     end
     event = read_yokogawa_event(filename, 'detectflank', detectflank, 'chanindx', chanindx, 'threshold', threshold);
     
+  case {'York_Instruments_hdf5'}
```
Comment (robertoostenveld) [inline fileio/ft_read_event.m:2241]: idem `yorkinstruments_hdf5`

File: fileio/ft_read_header.m [hunk 5]
```diff
@@ -2750,6 +2751,19 @@
   case 'video'
     hdr = read_video(filename);
     checkUniqueLabels = false;
+
+  case 'York_Instruments_hdf5'
```
Comment (robertoostenveld) [inline fileio/ft_read_header.m:2755]: idem `yorkinstruments_hdf5`

File: fileio/ft_read_headshape.m [hunk 6]
```diff
@@ -44,6 +44,7 @@
 %   '4d_*'
 %   'neuromag_*'
 %   'yokogawa_*'
+%   'York_Instruments_hdf5'
```
Comment (robertoostenveld) [inline fileio/ft_read_headshape.m:47]: idem `yorkinstruments_hdf5`

File: fileio/ft_read_sens.m [hunk 7]
```diff
@@ -524,7 +525,57 @@
     warning(ws); % revert to the previous warning state
     sens.label   = txtData{:,1};
     sens.elecpos = [txtData.Loc_X txtData.Loc_Y txtData.Loc_Z];
-    
+
+  case 'York_Instruments_hdf5'
```
Comment (robertoostenveld) [inline fileio/ft_read_sens.m:529]: idem `yorkinstruments_hdf5`

File: fileio/ft_read_sens.m [hunk 8]
```diff
@@ -24,6 +24,7 @@
 %   4d_pdf 4d_m4d 4d_xyz ctf_ds ctf_res4 itab_raw itab_mhd netmeg neuromag_fif
 %   neuromag_mne neuromag_mne_elec neuromag_mne_grad polhemus_fil polhemus_pos
 %   zebris_sfp spmeeg_mat eeglab_set localite_pos artinis_oxy3 artinis_oxyproj matlab
+%   York Instruments
```
Comment (robertoostenveld) [inline fileio/ft_read_sens.m:27]: please use the file format identifier `yorkinstruments_hdf5`

File: fileio/private/read_yi_hdf5_meta.m [hunk 9]
```diff
@@ -0,0 +1,78 @@
+function [info] = read_yi_hdf5_meta(datafile, acq_run)
```
Comment (robertoostenveld) [inline fileio/private/read_yi_hdf5_meta.m:1]: I suggest you rename this file to `read_yorkinstruments_hdf5_meta`.

File: forward/ft_senstype.m [hunk 10]
```diff
@@ -500,7 +500,7 @@
     case 'egi'
       type = any(strcmp(type, {'egi' 'egi32' 'egi64' 'egi128' 'egi256'}));
     case 'meg'
-      type = any(strcmp(type, {'meg' 'ctf' 'ctf64' 'ctf151' 'ctf275' 'ctf151_planar' 'ctf275_planar' 'neuromag' 'neuromag122' 'neuromag306' 'neuromag306_combined' 'bti' 'bti148' 'bti148_planar' 'bti248' 'bti248_planar' 'bti248grad' 'bti248grad_planar' 'yokogawa' 'yokogawa9' 'yokogawa160' 'yokogawa160_planar' 'yokogawa64' 'yokogawa64_planar' 'yokogawa440' 'itab' 'itab28' 'itab153' 'itab153_planar' 'babysquid' 'babysquid74' 'artenis123' 'magview'}));
+      type = any(strcmp(type, {'meg' 'ctf' 'ctf64' 'ctf151' 'ctf275' 'ctf151_planar' 'ctf275_planar' 'neuromag' 'neuromag122' 'neuromag306' 'neuromag306_combined' 'bti' 'bti148' 'bti148_planar' 'bti248' 'bti248_planar' 'bti248grad' 'bti248grad_planar' 'yokogawa' 'yokogawa9' 'yokogawa160' 'yokogawa160_planar' 'yokogawa64' 'yokogawa64_planar' 'yokogawa440' 'itab' 'itab28' 'itab153' 'itab153_planar' 'babysquid' 'babysquid74' 'artenis123' 'magview' 'York 4d'}));
```
Comment (robertoostenveld) [inline forward/ft_senstype.m:503]: it should be a single string, no spaces, no capitals. Would `bti248` not be better to describe the sensor array? In that case the template layouts and other 4d specific handling will automatically apply.

We could also make it `yorkinstruments248` if you consider the system sufficiently different from the `bti248`.

File: test/test_yorkinstruments.m [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+% MEM 3gb
+% WALLTIME 00:10:00
+% DEPENDENCY ft_filetype ft_read_header ft_read_data ft_read_event
+
+%%
+% this is where the data is installed on the DCCN compute cluster
+% and where the nightly regression testing is done
+
+filename = dccnpath('/home/common/matlab/fieldtrip/data/test/original/yi.meghdf5');
```
Comment (robertoostenveld) [inline test/test_yorkinstruments.m:11]: please update this to `/home/common/matlab/fieldtrip/data/test/original/meg/yorkinstruments/yi.meghdf5`

File: test/test_yorkinstruments.m [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+if length(ft_channelselection('MEG', hdr.label))~=246
+  error('did not select all MEG channels');
+end
+
+if ~ft_senstype(hdr, 'bti248')
+  error('hdr ~= bti248');
+end
+
+if ~strcmp(ft_filetype(filename), 'York_Instruments_hdf5')
```
Comment (robertoostenveld) [inline test/test_yorkinstruments.m:22]: should be updated in line with others changes

**Review by robertoostenveld — CHANGES_REQUESTED** [review]

**Comment by robertoostenveld** [comment]
If you want, you can already make the changes and commit them to your `RA-Neuro:YI_hdf5_reader` branch, that will ensure they end up here. Otherwise I can also make them once I review and test the code on my local computer (which won't be today).

File: fileio/ft_filetype.m (hunk 2 above)
Comment (RA-Neuro) [reply fileio/ft_filetype.m:1545]: That seems sensible, I have changed all instances in latest commit.

File: forward/ft_senstype.m (hunk 10 above)
Comment (RA-Neuro) [reply forward/ft_senstype.m:503]: I have chosen yorkinstruments248 as 4d system uses (I believe) ALS coordinate system while York Instruments switched to RAS (albeit for the same physical sensors). 
ft_senstype identifies the YI sensors as bti248 based on having the same set of channel names, but this is not problematic as the sensor types and units are the same for both systems.

**Comment by robertoostenveld** [comment]
could you check that you have pushed all your changes? From your comments it seems that you made more changes than I can see so far.,.

**Comment by RA-Neuro** [comment]
Sorry, changes should be present now.

**Comment by robertoostenveld** [comment]
Thanks, with some small code updates it all looks good. 

I have merged it with the master branch, and it will be included in the next [release](https://github.com/fieldtrip/fieldtrip/releases).

**Pull request merged**